In this program we can going to implement bagging technique to this model

In [1]:
# To plot all the graph within the note book 
%matplotlib inline

In [2]:
#importing all the required libraries 

import numpy as np
import pandas as pd 
from sklearn import metrics
from sklearn.tree import DecisionTreeClassifier
import seaborn as sns 

In [3]:
#loading the datset 

df=pd.read_csv("credit.csv")
df=pd.DataFrame(df)

In [4]:
#displaying dataset to know about independent and dependent variables 
df.head()

checking_balance  months_loan_duration credit_history               purpose  \
0           < 0 DM                     6       critical  furniture/appliances   
1       1 - 200 DM                    48           good  furniture/appliances   
2          unknown                    12       critical             education   
3           < 0 DM                    42           good  furniture/appliances   
4           < 0 DM                    24           poor                   car   

   amount savings_balance employment_duration  percent_of_income  \
0    1169         unknown           > 7 years                  4   
1    5951        < 100 DM         1 - 4 years                  2   
2    2096        < 100 DM         4 - 7 years                  2   
3    7882        < 100 DM         4 - 7 years                  2   
4    4870        < 100 DM         1 - 4 years                  3   

   years_at_residence  age other_credit housing  existing_loans_count  \
0                   4   67         none     own                     2   
1                   2   22         none     own                     1   
2                   3   49         none     own                     1   
3                   4   45         none   other                     1   
4                   4   53         none   other                     2   

         job  dependents phone default  
0    skilled           1   yes      no  
1    skilled           1    no     yes  
2  unskilled           2    no      no  
3    skilled           2    no      no  
4    skilled           2    no     yes

In [5]:
#by the above info we can notice that there are more number of non-numerical columns
# converting them into numerical type

for feature in df.columns:
    if df[feature].dtype=='object':
        df[feature]=pd.Categorical(df[feature]).codes
        

In [8]:
#splitting the datset into training and testing 

train_set=df.head(700)
test_set=df.tail(300)

train_labels=train_set.pop('default')
test_labels=test_set.pop('default')

In [9]:
dt_model=DecisionTreeClassifier(criterion='entropy')

In [10]:
dt_model.fit(train_set,train_labels)

DecisionTreeClassifier(criterion='entropy')

In [11]:
dt_model.score(test_set,test_labels)

0.6933333333333334

In [12]:
dt_model.score(train_set,train_labels)

1.0

we can observe that there is 100% accuracy on the training set, overfitting case, but for the test case it is 69%

USING ENSEMBLE TECHNIQUE BAGGING 

In [13]:
credit_labels=df.pop('default')

In [17]:
from sklearn.ensemble import BaggingClassifier
bgcl=BaggingClassifier(n_estimators=50,max_samples=0.8,oob_score=True) # by default oob_score will be false 
bgcl=bgcl.fit(df,credit_labels)
print(bgcl.oob_score_)

0.739


Inference : By using ensemble technique our model accuracy increased from 69% to 74%